src/VisualStudio/Xaml/Impl/Implementation/XamlProjectService.cs

```diff
@@ -83,54 +84,70 @@ public void TrackOpenDocument(string filePath)
-                var projectInfo = new VisualStudioProjectCreationInfo
+                if (!_xamlProjects.TryGetValue(hierarchy, out var project))
                 {
-                    Hierarchy = hierarchy,
-                    FilePath = hierarchy.TryGetProjectFilePath(),
-                    ProjectGuid = projectGuid
-                };
-
-                project = _visualStudioProjectFactory.CreateAndAddToWorkspace(name, StringConstants.XamlLanguageName, projectInfo);
-                _xamlProjects.Add(hierarchy, project);
-            }
-
-            if (!project.ContainsSourceFile(filePath))
-            {
-                project.AddSourceFile(filePath);
+                    if (!hierarchy.TryGetName(out var name))
+                    {
+                        return;
+                    }
+
+                    if (!hierarchy.TryGetGuidProperty(__VSHPROPID.VSHPROPID_ProjectIDGuid, out var projectGuid))
+                    {
+                        return;
+                    }
+
+                    var projectInfo = new VisualStudioProjectCreationInfo
+                    {
+                        Hierarchy = hierarchy,
+                        FilePath = hierarchy.TryGetProjectFilePath(),
+                        ProjectGuid = projectGuid
+                    };
+
+                    project = _visualStudioProjectFactory.CreateAndAddToWorkspace(name, StringConstants.XamlLanguageName, projectInfo);
+                    _xamlProjects.Add(hierarchy, project);
+                }
 
-                var documentId = _workspace.CurrentSolution.GetDocumentIdsWithFilePath(filePath).First(d => d.ProjectId == project.Id);
-                var document = _workspace.CurrentSolution.GetDocument(documentId)!;
-                var hasText = document.TryGetText(out var text);
-                if (!hasText || text?.Container.TryGetTextBuffer() == null)
+                if (!project.ContainsSourceFile(filePath))
                 {
-                    var docInfo = _rdt.GetDocumentInfo(docCookie);
-                    var textBuffer = _editorAdaptersFactory.GetDocumentBuffer(docInfo.DocData as IVsTextBuffer);
-                    var textContainer = textBuffer.AsTextContainer();
-                    _workspace.OnDocumentTextChanged(documentId, textContainer.CurrentText, PreservationMode.PreserveIdentity);
+                    project.AddSourceFile(filePath);
+
+                    var documentId = _workspace.CurrentSolution.GetDocumentIdsWithFilePath(filePath).Single(d => d.ProjectId == project.Id);
+                    var document = _workspace.CurrentSolution.GetDocument(documentId)!;
+                    var hasText = document.TryGetText(out var text);
+                    if (!hasText || text?.Container.TryGetTextBuffer() == null)
+                    {
+                        var docInfo = _rdt.GetDocumentInfo(docCookie);
+                        var textBuffer = _editorAdaptersFactory.GetDocumentBuffer(docInfo.DocData as IVsTextBuffer);
```

Comment:
Not calling the TryGetTextBufferFromDocData here?

Reply:
(the "as" doesn't make any sense here, because if that cast fails this will then throw an ArgumentNullException from GetDocumentBuffer.)

Reply:
Yeah, that would make more sense.  :-)

Reply:
FWIW I sent a pull request to the editor to null annotate the IVsEditorAdaptersFactoryService to catch bugs like this in the future.